RSE-366: Allow Dashboard Action Buttons

reneolivo opened the pull request:
## Overview
This PR allows extensions to add Action Buttons to the dashboard. This will help extensions like Awards to add custom action buttons to their dashboards.

## Before
![Screen Shot 2019-12-09 at 10 24 06 PM](https://user-images.githubusercontent.com/1642119/70489609-a613ef00-1ad2-11ea-97d6-09841972d532.png)

## After
![pr-after](https://user-images.githubusercontent.com/1642119/70489534-7369f680-1ad2-11ea-801c-c44cb4d499ef.gif)
**Note:** These are example buttons. If no extension configure their own buttons the dashboard should be shown as normal.

## Technical Details

### Action buttons

A new `DashboardActionButtons` provider has been created. This provider can be used by configuration files to add action buttons using the following structure:

```js
var actionButtons = [
  {
    buttonClass: 'btn btn-primary civicase__dashboard__action-btn',
    iconClass: 'add_circle',
    identifier: 'CustomButton',
    label: 'Custom Button',
    weight: 999
  }
];

DashboardActionButtonsProvider.addButtons(actionButtons);
```

* `buttonClass` is the class that will be applied to the button.
* `iconClass` is the Material Design icon's name.
* `identifier` is the name of a service that will handle the visibility and click events. The name of the actual service should follow this pattern: ***identifier**DashboardActionButton*, ext: `CustomButtonDashboardActionButton`.
* `label` is the text the user will see.
* `weight` is used to sort the list of buttons.

When requesting the list of buttons, after they have been configured, the provider will automatically append the corresponding service. If the service can't be found, a No Operation service is provided. This service has the expected methods every action button service should implement: `isVisible` and `clickHandler`, but nothing happens when they are called. The advantage of this is that we don't need to conditionally call methods if the service is not present.

Finally, the class for the "Add Case" button was changed from `civicase__dashboard__add-case` to `civicase__dashboard__action-btn`. The reason for this is that this class might be necessary to be referenced by buttons added using the provider. 

An extra `civicase__dashboard__action-btn--light` class was also added which displays an action button with white background.

### Add case

The add case button was also converted to an action button. The 3 HTML elements that were conditionally shown were reduced to a single action button service.

The button is always visible if the user has the `add cases` permission, but will behave differently when clicked. These are the rules:

* If a "New Case" webform url has been defined, it will redirect to this URL when the button is clicked. This can be defined on the civicase settings.
* If clicking the button from the dashboard of a particular case category, it will open the form only these case types. IE: when on the awards dashboard, it will only allow to create award case types.
* If not on a special dashboard, it will show the default case types.

### Settings

The following configuration options were moved from `civicase` to `civicase-base` and added as  Angular constants:

* `allowCaseLocks`
* `allowMultipleCaseClients`
* `defaultCaseCategory`
* `newCaseWebformClient`
* `newCaseWebformUrl`

The goal is to make testing easier by being able to mock the values at run time. All files referencing these values directly have been updated.

deb1990 commented on ang/civicase/dashboard/directives/dashboard.directive.html:
```diff
@@ -29,22 +29,32 @@ <h1 class="hidden" crm-page-title crm-document-title="myCasesOnly ? ts('CiviCase
              crm-popup-form-success="refresh()">
             <i class="material-icons">add_circle</i>{{ ts('Add Case') }}
           </a>
           <a ng-if="checkPerm('add cases') && newCaseWebformUrl"
```
As now we can add buttons using the newly added Provider. Please move the existing buttons to the same approach.
Also in the existing code, we have the "Add Case" button 3 times, all with different `ng-if` logic. 
So lets refactor it and have a single button, and all the ng-if should be moved to `isVisible` function logic.

PS: I thought I mentioned this in the ticket, as there was a discussion between myself and @tunbola . But looks like I forgot. Apologies.

reneolivo replied:
That is a great idea @deb1990

deb1990 commented on scss/components/_dashboard.scss:
```diff
@@ -33,6 +35,11 @@
     top: 2px;
   }
+
+  &--white {
```
Do we have a requirement to have this White button?

deb1990 replied:
If so, lets avoid having color names and follow semantic naming. maybe `--light`?

reneolivo replied:
@deb1990 this was requested by @jamienovick . I made a mistake and added it to another ticket.... but you can read it in the comments of RSE-301 😞

deb1990 replied:
Oh Okay. I just wanted to understand where the requirement is coming from. Lets still change the name of the class please.

reneolivo commented:
Notes for:

* 4cbfd4f
* 121a02e
* 3c8776a

They fix an issue found by QA. When I moved the code that exposes the configuration settings to civicase base I forgot to move a function that is also used in civicase to add an option to the action menu.

Fixed the issue by creating a helper class that includes all the functions related to "new case webform".

@deb1990 I also had to switch from `$location.url(...)` to `$window.location.href` because the former only replaces the hash location (ie: it only changes the angular page) and the former changes the whole URL, which is the expected result.

Also updated the setting's description because the form does not work using the example mentioned in the previous description.

Adding a gift with the AFTER for this particular feature:

![Screen Shot 2019-12-18 at 6 07 25 PM](https://user-images.githubusercontent.com/1642119/71127341-4fd53900-21c1-11ea-96da-56aaf7bc9186.png)

![gif](https://user-images.githubusercontent.com/1642119/71127703-eb66a980-21c1-11ea-94f3-bfb3cd9bbf0f.gif)

deb1990 commented on ang/civicase/dashboard/services/add-case-dashboard-action-button.service.js:
```diff
@@ -7,11 +7,11 @@
    * Dashboard Action Button for adding new cases.
    *
    * @param {object} $location the location service.
```
We should remove `$location` as its not used anymore?

reneolivo replied:
The `$location` service is still used to determine on which dashboard we currently are (cases, awards, prospects, etc.):
https://github.com/compucorp/uk.co.compucorp.civicase/blob/ae0ef4a0d3f30a9efa474f89dad4877958271b4b/ang/civicase/dashboard/services/add-case-dashboard-action-button.service.js#L36-L40

deb1990 commented on ang/civicase/dashboard/services/add-case-dashboard-action-button.service.js:
```diff
@@ -7,11 +7,11 @@
    * @param {string} newCaseWebformUrl the new case web form url configuration value.
    */
-  function AddCaseDashboardActionButton ($location, $rootScope, defaultCaseCategory,
+  function AddCaseDashboardActionButton ($location, $window, defaultCaseCategory,
```
We should remove `$location`?

reneolivo replied:
See comment above.

deb1990 commented on ang/test/civicase/dashboard/services/add-case-dashboard-action-button.service.spec.js:
```diff
@@ -2,14 +2,20 @@
 (($, loadForm, getCrmUrl) => {
   describe('AddCaseDashboardActionButton', () => {
-    let $rootScope, $location, AddCaseDashboardActionButton, defaultCaseCategory,
+    let $location, $window, AddCaseDashboardActionButton, defaultCaseCategory,
```
We should remove `$location`?

reneolivo replied:
Same as above.

tunbola commented on CRM/Civicase/Helper/NewCaseWebform.php:
```diff
@@ -0,0 +1,52 @@
+  /**
+   * Retrieve civicase webform url.
+   */
+  public static function appendUrlToOptions(&$options) {
```
This function adds the webformURL and the webformClient to the options array. I think a better name will be 
`addWebformDataToOptions` or something better, we should change the function description to reflect that also.

tunbola commented on templates/CRM/Civicase/Admin/Form/Settings.tpl:
```diff
@@ -13,7 +13,7 @@
   <td class="label">{$form.civicaseWebformUrl.label}</td>
   <td>{$form.civicaseWebformUrl.html}<br />
-    <span class="description">{ts}A Webform url e.g node/233{/ts}</span>
+    <span class="description">{ts}A Webform url e.g /node/233{/ts}</span>
```
Are we sure we need to add a forward slash before the webform URL? tbh, not sure myself, just double checking with you.

reneolivo replied:
Yes @tunbola, this is what happens in different scenarios taking into consideration we currently are in `http://civicase.local/civicrm/case/a/#/case` (case dashboard):

* if the webform url value is `node/233` it will try to go to `http://civicase.local/civicrm/case/node/233`.
* if the webform url value is `/node/233` it will go to `http://civicase.local/node/233`.
* if the url value is `http://example.com` it will go to `http://example.com`.

So if we just add `node/233` it will go to a place that does not even exist because it will try to solve a relative path.

tunbola commented on CRM/Civicase/Helper/NewCaseWebform.php:
```diff
@@ -8,7 +8,7 @@ class CRM_Civicase_Helper_NewCaseWebform {
    * Retrieve civicase webform url.
    */
-  public static function appendUrlToOptions(&$options) {
+  public static function addWebformDataToOptions(&$options) {
```
Can we change the function description to reflect the name. I.e the function adds web form data to options array.
